### ISSUE TITLE
Editor: change sprite storage in the game project

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1281]
## Problem

For historical reasons sprites in AGS game project are kept compiled in a single file (aka "sprite file"). Editor saves them there on import and reads them back from there when having to use these sprites displaying a game scene. When compiling the game this file is simply added to the data package as-is, and this is it's only benefit.
But this has a number of disadvantages:
* Reliant on editor (user cannot modify sprite collection by hand, has to use editor or another tool);
* Vulnerable to disk errors (a single file corruption may make entire sprite collection unreadable);
* Not friendly to source control tools (especially if sprites are compressed), and therefore nearly impossible to delegate resource management in a team.
See also: #468.

## Solution summary

* Game project has sprites as separate normal image files in the project folder, or rather dedicated subfolder.
* When importing a sprite editor copies it to the folder, unless it's already there. Optionally editor may be made to import sprites automatically if new ones are found in the folder.
* The import settings are kept as they currently are (tile support, etc) - changing that is not a part of this task (not sure if there's any need either).
* Editor has its own sprite manager that loads sprites up from these image files when they are necessary (e.g. for displaying a game scene) and unloads when no longer necessary.
* When compiling the game Editor saves a classic sprite file by copying corresponding image data from the image files. From this point the process is similar to pre-existing one.

Questions and further suggestions follow...

### Project storage

Need to define the subfolder to store sprites. "Sprites"?

Personally I doubt if subfolders within "Sprites" should correspond to the folders in the editor's sprite manager. IMO more no than yes. The imported sprites do not 1:1 correspond to the source images, user may import multiple sprites from one image file, even from one tile rect with different settings. Same image may be imported as multiple placeholders too.
It may be imagined that at some point editor will be able to display both source images and imported sprites in separate windows - as separate entities.

### Importing sprites

On a sprite import editor checks whether the source image is withing "Sprites" project folder, if not then copies it there.
If we support subfolders, images cloned by the editor should probably go into a special "unorganized" subfolder.
Editor will also have to prevent file name conflicts.

Alternatively we may restrict sprite import to only files that were placed into "Sprite" folder by user.
Pros: user will have full control on how files are called, there's no need to guess how and where to clone files. Possibly less confusion about how import works.
Cons: possibly extra work for users if they import from other locations.

What to do with "import from clipboard" option? If supported then it must create a new source image file in the project folder.

### Upgrading from older project

Existing AGS project has:
* sprite information in Game.agf,
* source images somewhere on disk (may be within or outside of the project folder),
* saved sprites in the compiled sprite file.

Let's suppose that the priority should be given to importing original images whenever possible. If they cannot be found, then we may optionally export sprites from the compiled sprite file and save them in "Sprites" folder.

If we are reusing existing image sources, what to do if there's already a "Sprites" folder created by user and containing some data?

If we are exporting from the spritefile, we will have to deal with possible duplicate image sources and sprites cut out from the middle of an image. E.g. there may be multiple sprites referencing one source image. In such case should we try to reconstruct original image judging by the sprite tile offsets & sizes? Or should we export each sprite to its own image file and adjust source references and import settings?
Should the program ask for permission to export from the spritefile to let user opportunity to restore original images, or just warn and go on with it?

### Sprites in the Editor's memory

Hypothetically, Editor may reuse existing sprite cache in C++ code with some adjustments. If it's preferable to move away from using C++ code internally in the Editor, then we would have to write another sprite cache or manager for the Editor in C#.

### Editor drawing sprites

If we remake sprite cache into a C# cache/manager, then there very likely is a need to **remake remaining game content drawing into C# code** as well. This task may be stated in a separate ticket.

### Compiling sprite file

The existing C++ code relies on sprites stored in a native C++ sprite cache object, which is going to be inconvenient.

The options here are:
* Separate spritefile handling into another C++ class, which may accept sprite information in a different, convenient form.
* Write a standalone tool based on existing C++ code, and pass source sprite information in a convenient form.
* Write a spritefile writing code in C#, whether internally in the Editor or also as a stand-alone tool.

Writing stand-alone tool may be picked out to a separate ticket.

**Comment by ivan-mogilko** [comment]
@selmiak the original idea was to have actual source sprites inside a project's subfolder, where user can work, directly editing these sprites, and editor would catch up these changes automatically.

You are speaking of a "sprite cache", which seems more like an optional internal feature for the editor program's convenience or perfomance. I think we should separate these two things.

**Comment by persn** [comment]
The data for all the sprites and how they will show as folders in the editor are found inside the Game.agf file. Anything not added there will not show.

![image](https://user-images.githubusercontent.com/1510803/137981173-d379cdc8-4ffb-411f-bf25-b1c9c51920c6.png)

**Comment by persn** [comment]
I got started on a branch for this while I was awaiting feedback on the open room format PR a while back, but it has been collecting dust for a while now. I didn't get super far, it takes the sprites through the native poxy, dumps it to disk, and there's a sprite drawing that draws using sprite from disk that needs more work.

I might do more on it if I can, but just to be safe, I'll try to rebase it on ags4 branch and post it here this weekend so anyone else could use it for reference if I don't have time

**Comment by ivan-mogilko** [comment]
@persn may be relevant, recently I wrote a managed SpriteFileWriter wrapper meant to recompile spritefile by importing sprites, perhaps could be seen as a reference: #1594

**Comment by persn** [comment]
> @persn may be relevant, recently I wrote a managed SpriteFileWriter wrapper meant to reconstruct spritefile by reimporting all sprites from sources, perhaps could be seen as a reference for spritefile compiling from the project assets: https://github.com/adventuregamestudio/ags/pull/1594

I imagine one of the end goals will be to have the ability to gitignore the sprite file (this should be possible now with .crm files), so that would be absolutely relevant

**Comment by persn** [comment]
> I got started on a branch for this 

https://github.com/persn/ags/tree/feature/open-sprites

I'll keep it up to date if I make any more changes

**Comment by ericoporto** [comment]
Talked a bit with @persn, and he mentioned that his branch was showing two PRs were necessary.

- **1st PR**: Consolidate sprite code using the [`ISpriteController`](https://github.com/adventuregamestudio/ags/blob/master/Editor/AGS.Types/Interfaces/ISpriteController.cs) interface, similar to `IRoomController` for open room format. Today, multiple rendering methods exist, including an unrelated custom drawing routine in the editor plugin interface, these have to be unified in a way to avoid code duplication.
- **2nd PR**: Replace the interface implementation with code that reads from the open format.

**Comment by ivan-mogilko** [comment]
I was going to write this for a while, but could not find spare time.

I think that besides the above, there's another change that would be necessary, and unfortunately it was not mentioned in the ticket (as far as I can see). This is related to test-running the game.

The spritefile compilation may take some time, if there are alot of sprites. Right now it's being updated as you import sprites and save the project in the Editor, and the long wait time only happens when you save after adding or changing sprites. But this process copies parts of the previous file to speed things up a bit.

If we change this to have sprites on disk, and spritefile built only during game compilation, that raises the question about how to update a spritefile. Because if the spritefile is fully recreated each time by doing a import from separate files (let alone doing tiled import etc, - I am not fully certain how optimized the storage would be), then this process will become slower than before.

Therefore, the Editor would either have to keep a record of what has changed since the last spritefile recompilation, track the time of modification of a spritefile, etc.
Or there will have to be some kind of an *intermediate form of sprite storage*, between the source image files in the folder, and compiled spritefile, that would speed up gathering final game file.

In the second case, for the optimal workflow, it would be desired to also let Engine *use that intermediate form*, in such case the full recompilation would not be necessary during the test run.

There is this ticket #702, which somehow got almost 50 comments in the discussion, and for the most part we discussed various options of solving similar problem for a different reason. If I recall right, the discussed methods may be summarized into following variants:
1. Store the results of import and conversion to raw bitmap (or whatever engine can load) in separate files. That is - not true BMP files with BM header and such, but raw chunks of pixel data in the same format as these chunks are written in the spritefile (may even be compressed).
2. Have a "dirty" spritefile which stores both old "deleted" sprites and new ones and replacements to old ones. Don't delete old sprite data, only append. Use index table to know where are the true sprites in this file. Every now and then recompile the file to remove deprecated chunks.
3. An extended variant of p2: store a dirty spritefile in multiple partitions. When you change something, only append to the respective partition, or start a new one.

The idea here is that besides the editor, that should handle this intermediate storage, the Engine should also be taught to read from it. Since now we have a completely separate SpriteFile class that streams from a file, and SpriteCache itself does not care how the sprite data is stored on disk, it will be possible to have different implementations of the same SpriteFile interface, including one reading from separate files, etc. They likely will work slower than streaming from 1 file, but with a well-sized cache that may not be a big problem. The choice of implementation may depend on a game setting or a command line arg.

Above is a problem overview, I think we would need a dedicated ticket for this. Note that likely this may be worked on in parallel to the other subtasks.

**Comment by ericoporto** [comment]
Because we can recreate the spritefile from source files nowadays, I still don't quite grasp what this open format adds :/. Additionally, as sprites are sequential numbered items, a merge of branches will always entail manual sprite index adjustment or following practices from cisbetter tutorial.

**Comment by ivan-mogilko** [comment]
> Because we can recreate the spritefile from source files nowadays, I still don't quite grasp what this open format adds 

Our goal is to have fully open project source, where everything is stored in a human-readable format, without using "magic packages", where:
* a user may view and edit things without opening editor,
* the editor may react to any part of source changing externally and update the preview automatically, without user having to do "reimport from source" command all the time,
* things may be easily merged together, and its easy to see the history of changes.

> Additionally, as sprites are sequential numbered items, a merge of branches will always entail manual sprite index adjustment

Yes, this problem still persist, and have to be addressed separately. I might suggest discussing this on forums, to see if people have some interesting ideas.

One idea that I once had is to implement a workspace setting that defines a range of sprite IDs to use when importing new sprites. Then each team member could define different range for themselves, and thus avoid conflicts.

**Comment by persn** [comment]
> Talked a bit with @persn, and he mentioned that his branch was showing two PRs were necessary.
> 
> * **1st PR**: Consolidate sprite code using the [`ISpriteController`](https://github.com/adventuregamestudio/ags/blob/master/Editor/AGS.Types/Interfaces/ISpriteController.cs) interface, similar to `IRoomController` for open room format. Today, multiple rendering methods exist, including an unrelated custom drawing routine in the editor plugin interface, these have to be unified in a way to avoid code duplication.
> * **2nd PR**: Replace the interface implementation with code that reads from the open format.

Would PR 1 be suitable for the master branch? Or the ags4 branch?

**Comment by ivan-mogilko** [comment]
> Would PR 1 be suitable for the master branch? Or the ags4 branch?

Could you elaborate on which changes will this include?

**Comment by persn** [comment]
> Could you elaborate on which changes will this include? Overall 3.* branch is today meant primarily for improving existing features (performance, convenience, cleaning the code), larger changes should go into ags4.

I don't think I can explain it anymore plainly than what's already written. It's probably easy enough to rebase, so it would make sense to just base the branch off master, and then rebase on request after a PR review if the changes got out of scope
